### ISSUE TITLE
Account spoofing

### ISSUE THREAD

**Issue by 0x010C** [irods/irods_auth_plugin_pam_interactive#18]
Good morning,

I am currently testing this plugin (thanks for that!) to delegate the authentication of my users to an SSO (keycloak) via the OIDC protocol.

My tests worked well with [pam_deviceauthgrant](https://github.com/ll4strw/pam_deviceauthgrant) (but also with [PAM-OIDC-Device-Flow](https://gitlab.kodden.nl/harry/PAM-OIDC-Device-Flow)), however, I came across a strange behavior where I was able to log into another user's account (see test-case 4 below).

Here are my configuration files:

## Server-side configuration
**/etc/irods/server_config.json**
```javascript
    ...
    "plugin_configuration": {
        "authentication": {
                "pam_interactive" : {
                        "password_min_time": 3600,
                        "password_max_time": 7200
                }
        },
        ...
    },
    ...
```

**/etc/pam.d/irods**
```sh
auth    required        pam_nologin.so
auth    sufficient      /usr/local/lib64/security/pam_deviceauthgrant.so debug
auth    required        pam_deny.so
account sufficient      /usr/local/lib64/security/pam_deviceauthgrant.so
```

## Client-side configuration
**~/.irods/irods_environment.json**
```javascript
{
    "irods_authentication_scheme": "pam_interactive",
    "irods_host": "my.irods.domain.name",
    "irods_port": 1247,
    "irods_user_name": "userA",
    "irods_zone_name": "myZone"
}
```

## What I have done
<details><summary>Successful test-cases (for information and context)</summary>
<p>

### Test-case 1
Set "userA" in irods_environment.json.
When prompted for login by pam_interactive, type "userA".
When the SSO login-page opens, log-in as "userA".
```shell
$ iinit
login: userA
Login at https://my.sso.domain.name/realms/test/device?user_code=ABCD-EFGH
After login press Enter 
$ ils
/myZone/home/userA:
  testfile.json
``` 
**Expected behaviour:** Be loged-in as userA
**Result:** :heavy_check_mark: Loged-in as userA

### Test-case 2
Set "userA" in irods_environment.json.
When prompted for login by pam_interactive, type "userB".
When the SSO login-page opens, log-in as "userA".
```shell
$ iinit
login: userB
Login at https://my.sso.domain.name/realms/test/device?user_code=ABCD-EFGH
After login press Enter 

Level 0: Error occurred while authenticating user [userA] [CAT_INVALID_AUTHENTICATION: authentication flow completed without success
] [ec=-826000]
``` 
**Expected behaviour:** Login error
**Result:** :heavy_check_mark: Login error

### Test-case 3
Set "userA" in irods_environment.json.
When prompted for login by pam_interactive, type "userA".
When the SSO login-page opens, log-in as "userB".
```shell
$ iinit
login: userA
Login at https://my.sso.domain.name/realms/test/device?user_code=ABCD-EFGH
After login press Enter 

Level 0: Error occurred while authenticating user [userA] [CAT_INVALID_AUTHENTICATION: authentication flow completed without success
] [ec=-826000]
``` 
**Expected behaviour:** Login error
**Result:** :heavy_check_mark: Login error

</p>
</details> 

### Test-case 4
Set "userA" in irods_environment.json.
When prompted for login by pam_interactive, type "userB".
When the SSO login-page opens, log-in as "userB".
```shell
$ iinit
login: userB
Login at https://my.sso.domain.name/realms/test/device?user_code=ABCD-EFGH
After login press Enter 
$ ils
/myZone/home/userA:
  testfile.json
``` 
**Expected behaviour:** Be loged-in as userB or login error
**Result:** :red_circle:  Loged-in as userA

In other words, I logged into userA account using userB credentials.

----

I'm not sure if it's a configuration error on my part (in this case sorry for the inconvenience) or a pam_interactive bug, but in the latter case it's a major security issue.

Thanks and best regards

**Comment by alanking** [comment]
Thanks for bringing this to our attention. This is indeed a bug in pam_interactive, *not* a configuration error on your part.

The fix would be to ensure that the username being fed to PAM matches the username which is being authenticated with iRODS.

We will try to fix it as soon as we can. As a reminder, this plugin has not been officially released, so it may be slow going for the severity level.

@ll4strw - Just giving you a notification of this issue. We do not think this is a problem with the deviceauthgrant module, but want you to be aware of its existence.

**Comment by 0x010C** [comment]
@ll4strw I've compiled your commit and my test-cases are working as expected now, thanks!

**Comment by trel** [comment]
I haven't run it - but... this solution still prompts for the username once?

So, in Test-case 4... 'userB' is given as the answer to the prompt... and then fails to get past the plugin because of the new confirmation check?  Because 'userB' doesn't match the environment's 'userA'?

**Comment by 0x010C** [comment]
No, there is no prompts for the username anymore. In my case (using [pam_deviceauthgrant](https://github.com/ll4strw/pam_deviceauthgrant) it looks like this now:
```shell
$ iinit
Login at https://my.sso.domain.name/realms/test/device?user_code=ABCD-EFGH
After login press Enter 
``` 

From what I see there is only two cases left :
- The username in irods_environment.json **matches** the username used on the SSO side
- The username in irods_environment.json **does not match** the username used on the SSO side

**Comment by trel** [comment]
Oh, even better.   Thanks for the confirmation!

**Comment by trel** [comment]
Got it.   I had missed the if/else case upon inspection.

**Comment by alanking** [comment]
#19 has been merged and has been confirmed to resolve the issue, so I am going to close this. Thanks @0x010C for the report, and @ll4strw for the fix!
